framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,693 @@
+     */
+    private static @Nullable ATypeElement getNextArrayLevel(@Nullable ATypeElement e) {
+        if (e == null) {
+            return null;
+        }
+
+        for (Map.Entry<InnerTypeLocation, ATypeElement> ite : e.innerTypes.entrySet()) {
+            InnerTypeLocation loc = ite.getKey();
+            if (loc.location.contains(TypePathEntry.ARRAY)) {
+                return ite.getValue();
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Formats a single formal parameter declaration.
+     *
+     * @param param the AFieldWrapper that represents the parameter
+     * @param parameterName the name of the parameter to display in the stub file. Stub files
+     *     disregard formal parameter names, so this is aesthetic in almost all cases. The exception
+     *     is the receiver parameter, whose name must be "this".
+     * @param basename the type name to use for the receiver parameter. Only used when the previous
+     *     argument is exactly the String "this".
+     * @return the formatted formal parameter, as if it were written in Java source code
+     */
+    private static String formatParameter(
+            AFieldWrapper param, String parameterName, String basename) {
+        return formatAFieldImpl(param, parameterName, basename);
+    }
+
+    /**
+     * Formats a field declaration or formal parameter so that it can be printed in a stub.
+     *
+     * <p>This method does not add a trailing semicolon or comma.
+     *
+     * <p>Usually, {@link #formatParameter(AFieldWrapper, String, String)} should be called to
+     * format method parameters, and {@link #printField(AFieldWrapper, String, PrintWriter, String)}
+     * should be called to print field declarations. Both use this method as their underlying
+     * implementation.
+     *
+     * @param aField the field declaration or formal parameter declaration to format; should not
+     *     represent a local variable
+     * @param fieldName the name to use for the declaration in the stub file. This doesn't matter
+     *     for parameters (except the "this" receiver parameter), but must be correct for fields.
+     * @param className the simple name of the enclosing class. This is only used for printing the
+     *     type of an explicit receiver parameter (i.e., a parameter named "this").
+     * @return a String suitable to print in a stub file
+     */
+    private static String formatAFieldImpl(
+            AFieldWrapper aField, String fieldName, String className) {
+        if ("this".equals(fieldName)) {
+            return formatType(aField.getTheField().type, null, className) + fieldName;
+        } else {
+            return formatType(aField.getTheField().type, aField.getType()) + fieldName;
+        }
+    }
+
+    /**
+     * Formats the given type for printing in Java source code.
+     *
+     * @param type the scene-lib representation of the type, or null if only the bare type is to be
```

Comment:
What is a "bare type"?  Is it an unannotated type?  Same question applies at other uses of this term.